nomos-da/network/core/src/swarm/common/monitor.rs

```diff
@@ -0,0 +1,387 @@
+// crates
+use fixed::types::U57F7;
+use libp2p::PeerId;
+use serde::{Deserialize, Serialize};
+// internal
+use crate::maintenance::monitor::{ConnectionMonitor, ConnectionMonitorOutput, PeerStatus};
+use crate::protocols::dispersal::executor::behaviour::{
+    DispersalError as ExecutorDispersalError, DispersalExecutorEvent,
+};
+use crate::protocols::dispersal::validator::behaviour::{
+    DispersalError as ValidatorDispersalError, DispersalEvent as DispersalValidatorEvent,
+};
+use crate::protocols::replication::behaviour::{ReplicationError, ReplicationEvent};
+use crate::protocols::sampling::behaviour::{SamplingError, SamplingEvent};
+
+pub enum MonitorEvent {
+    ExecutorDispersal(ExecutorDispersalError),
+    ValidatorDispersal(ValidatorDispersalError),
+    Replication(ReplicationError),
+    Sampling(SamplingError),
+    Noop,
+}
+
+impl MonitorEvent {
+    pub fn peer_id(&self) -> Option<&PeerId> {
+        match self {
+            MonitorEvent::ExecutorDispersal(dispersal_error) => dispersal_error.peer_id(),
+            MonitorEvent::ValidatorDispersal(dispersal_error) => dispersal_error.peer_id(),
+            MonitorEvent::Replication(replication_error) => replication_error.peer_id(),
+            MonitorEvent::Sampling(sampling_error) => sampling_error.peer_id(),
+            MonitorEvent::Noop => None,
+        }
+    }
+}
+
+impl From<&DispersalExecutorEvent> for MonitorEvent {
+    fn from(event: &DispersalExecutorEvent) -> Self {
+        match event {
+            DispersalExecutorEvent::DispersalSuccess { .. } => MonitorEvent::Noop,
+            DispersalExecutorEvent::DispersalError { error } => {
+                MonitorEvent::ExecutorDispersal(error.clone())
+            }
+        }
+    }
+}
+
+impl From<&DispersalValidatorEvent> for MonitorEvent {
+    fn from(event: &DispersalValidatorEvent) -> Self {
+        match event {
+            DispersalValidatorEvent::IncomingMessage { .. } => MonitorEvent::Noop,
+            DispersalValidatorEvent::DispersalError { error } => {
+                MonitorEvent::ValidatorDispersal(error.clone())
+            }
+        }
+    }
+}
+
+impl From<&ReplicationEvent> for MonitorEvent {
+    fn from(event: &ReplicationEvent) -> Self {
+        match event {
+            ReplicationEvent::IncomingMessage { .. } => MonitorEvent::Noop,
+            ReplicationEvent::ReplicationError { error } => {
+                MonitorEvent::Replication(error.clone())
+            }
+        }
+    }
+}
+
+impl From<&SamplingEvent> for MonitorEvent {
+    fn from(event: &SamplingEvent) -> Self {
+        match event {
+            SamplingEvent::SamplingSuccess { .. } => MonitorEvent::Noop,
+            SamplingEvent::IncomingSample { .. } => MonitorEvent::Noop,
+            SamplingEvent::SamplingError { error } => MonitorEvent::Sampling(error.clone()),
+        }
+    }
+}
+
+#[derive(Default, Debug)]
+pub struct PeerStats {
+    // Failure rate tracking (sliding window)
```

Comment:
why sliding window? I don't understand the comment

Reply:
Added a comment to the PeerStats struct and explanation to it's field types